aggregator_core/src/datastore.rs

```diff
@@ -1491,21 +1519,59 @@ impl<C: Clock> Transaction<'_, C> {
                       AND client_reports.client_timestamp < UPPER($2::TSRANGE)
                       AND client_reports.client_timestamp >= $3
                       AND client_reports.aggregation_started = FALSE
-                ) AS unaggregated_report_exists",
+                ) AS never_aggregated_report_exists",
             )
             .await?;
-        let row = self
+        let never_aggregated_report_exists = self
             .query_one(
-                &stmt,
+                &fast_path_stmt,
                 &[
                     /* task_id */ &task_info.pkey,
                     /* batch_interval */ &SqlInterval::from(batch_interval),
                     /* threshold */
                     &task_info.report_expiry_threshold(&self.clock.now().as_naive_date_time()?)?,
                 ],
             )
+            .await?
+            .get("never_aggregated_report_exists");
+
+        // TODO(timg): are we OK with runtime type check?
+        if never_aggregated_report_exists || TypeId::of::<AP>() == TypeId::of::<()>() {
+            return Ok(never_aggregated_report_exists);
+        }
+
+        // Slow path: aggregation parameter is non-trivial and we did not find any reports in the
+        // interval that have never been aggregated. We check whether there exist any unfinished
+        // aggregation jobs that match our aggregation parameter.
```

Comment:
I think you're right. I suspect I can delete this slow path altogether, which means this method won't need to be specialized based on aggregation parameter at all.

Reply:
I don't think we can delete this slow path: what this method needs to check is if there are any reports in the batch (identified by a time interval) that have not yet been associated to an aggregation job.

In the single-collection case, this is effectively checking that we aren't leaving any reports unaggregated. In the multi-collection case, this is additionally necessary to ensure we are aggregating the same set of reports for each collection of the same batch, which is necessary[1] for privacy.

One way to handle this would be to generate the explicit report-to-batch mapping discussed in Slack, and modify this check to use that mapping. Then we can get rid of the fast path for the multi-collection case, or perhaps even use the report-to-batch mapping for both (which would be a nice simplification, if it's not too unperformant). 

Note that this check would eventually need to be extended to the fixed-size case, too: right now we don't need a check at all since reports are associated with batches at the point of being associated with an aggregation param, so there is no possibility of a fixed-size batch having reports around that aren't associated with a batch. That will no longer be true once in a multi-collection world.

[1] unless something-something-DP? But I'm (personally) not comfortable relying on DP to cover for implementation bugs, if we can avoid it -- leaking one report, even noised, is still leaking a lot more information than _not_ leaking that report.

Reply:
In the time interval case, we already get the guarantee that the same set of reports gets collected each time, because once the interval is collected, the upload handler won't allow new reports in that interval. So we don't need any further checks at collect time.

I agree more is needed here to handle the fixed size case, but in this PR I only want to enable time interval multi-collection.

Reply:
The check being discussed is for reports being included in an aggregation job, rather than for report uploads. That is, this is a check for when we can step the collection job -- uploads are orthogonal to this problem (and are indeed OK for the reason you describe).

Without a check that all reports in the batch have been attached to an aggregation job, the specific problem that can occur is:
1) Some set of reports is uploaded to a batch B. (For purposes of this description, this set won't change over time due to new uploads.)
2) A collection request for batch B arrives with aggregation parameter A_1. We start generating aggregations for it.
3) Another collection request for batch B arrives with aggregation parameter A_2. We start generating aggregations for it.
4) The collection job driver attempts to step both collection jobs. Since there is no check as to whether all reports in the batch have started aggregation, both collection jobs are successfully stepped. The sets of aggregated reports will both be subsets of `B`, but will otherwise have no relation -- they may be identical, one may be a subset of the other, they may be disjoint, etc.

We need some way to ensure that for each collection, all reports in the batch have started aggregation before the collection job can be stepped. (We also have to check that all the aggregation jobs which were started have terminated, but that check does not need to change, I think.)

Reply:
I'm not sure that the upload handler's check would entirely prevent that, since we're using REPEATABLE READ isolation level. An upload transaction could plausibly have a stale view of collection jobs and still commit. Rather, I think what's load bearing in the single collection case is that collection flips the state of all `batch_aggregation` rows, and aggregation job creation would conflict by bumping a counter in one of the rows.